src/databricks/labs/ucx/azure/locations.py

```diff
@@ -0,0 +1,170 @@
+
+        return cls(ws, hms_locations, resource_permissions, azurerm)
+
+    def _app_id_credential_name_mapping(self) -> tuple[dict[str, str], dict[str, str]]:
+        # list all storage credentials.
+        # generate the managed identity/service principal application id to credential name mapping.
+        # return one mapping for all non read-only credentials and one mapping for all read-only credentials
+        # TODO: considering put this logic into the StorageCredentialManager
+        app_id_mapping_write = {}
+        app_id_mapping_read = {}
+        all_credentials = self._ws.storage_credentials.list(max_results=0)
+        for credential in all_credentials:
+            # cannot have none credential name, it's required for external location
+            if not credential.name:
+                continue
+            # if service principal based credential, use service principal's application_id directly
+            if credential.azure_service_principal:
+                if not credential.read_only:
+                    app_id_mapping_write[credential.azure_service_principal.application_id] = credential.name
+                    continue
+                if credential.read_only:
+                    app_id_mapping_read[credential.azure_service_principal.application_id] = credential.name
+            # if managed identity based credential, fetch the application_id of the managed identity
+            if credential.azure_managed_identity:
+                application_id = self._azurerm.managed_identity_client_id(
+                    credential.azure_managed_identity.access_connector_id,
+                    credential.azure_managed_identity.managed_identity_id,
+                )
+                if not application_id:
+                    continue
+                if not credential.read_only:
+                    app_id_mapping_write[application_id] = credential.name
+                    continue
+                if credential.read_only:
+                    app_id_mapping_read[application_id] = credential.name
+
+        return app_id_mapping_write, app_id_mapping_read
+
+    def _prefix_credential_name_mapping(self) -> tuple[dict[str, str], dict[str, str]]:
+        # get managed identity/service principal's application id to storage credential name mapping
+        # for all non read-only and read-only credentials
+        app_id_mapping_write, app_id_mapping_read = self._app_id_credential_name_mapping()
+
+        # use the application id to storage credential name mapping to create prefix to storage credential name mapping
+        prefix_mapping_write = {}
+        prefix_mapping_read = {}
+        for permission_mapping in self._resource_permissions.load():
+            if permission_mapping.client_id in app_id_mapping_write:
+                prefix_mapping_write[permission_mapping.prefix] = app_id_mapping_write[permission_mapping.client_id]
```

Comment:
This data structure is incorrect, because if prefix has multiple readers or writers - the last one overrides the previous.

Correct data structure is dict[str,set[str]]

Reply:
wWell this is intentional. We cannot have multiple UC external locations pointing to same prefix, so when multiple credentials have access to one same prefix, we can only pick one credential to create the external location. So the logic is first try pick one that has write access, if not pick one that has read access. So the data structure does not need to contain all readers and writers